kong/plugins/opentelemetry/handler.lua

```diff
@@ -147,7 +131,39 @@ function OpenTelemetryHandler:access(conf)
   -- Set the sampled flag for the outgoing header's span
   injected_parent_span.should_sample = sampled
 
-  propagation_set(conf.header_type, header_type, injected_parent_span, "w3c")
+  extract_ctx.trace_id      = injected_parent_span.trace_id
+  extract_ctx.span_id       = injected_parent_span.span_id
+  extract_ctx.should_sample = injected_parent_span.should_sample
+  extract_ctx.parent_id     = injected_parent_span.parent_id
+
+  -- return the injected ctx (data to be injected with outgoing tracing headers)
+  return extract_ctx
+end
+
+function OpenTelemetryHandler:access(conf)
+  local root_span = ngx.ctx.KONG_SPANS and ngx.ctx.KONG_SPANS[1]
+
+  -- get the global tracer when available, or instantiate a new one
+  local tracer = kong.tracing.name == "noop" and kong.tracing.new("otel")
+                 or kong.tracing
+
+  -- make propagation work with tracing disabled
+  if not root_span then
+    root_span = tracer.start_span("root")
+    root_span:set_attribute("kong.propagation_only", true)
+
+    -- since tracing is disabled, turn off sampling entirely for this trace
+    kong.ctx.plugin.should_sample = false
+  end
+
+  local injected_parent_span = tracing_context.get_unlinked_span("balancer") or root_span
+
+  propagation.propagate(
```

Comment:
2 questions about this design:
1. This makes the plugin really a light warp over the core logic. Does it make more sense to have a single "tracing plugin" and make it support all different types of tracing systems?
2. What if we enable more than 1 tracing plugin and they invoke the `propagate()` more than once, with different configurations?

Reply:
1. the idea here is to separate tracing from propagation to let _tracing_ plugins handle the update and report of the tracing context and then call the propagation logic which is supposed to be common. The idea of having a single plugin for propagation crossed our minds but there are differences to the ways tracing plugins handle the information they extract/inject from/to headers, for example [Zipkin](https://github.com/openzipkin/b3-propagation?tab=readme-ov-file#why-is-parentspanid-propagated) and [OpenTelemetry](https://opentelemetry.io/docs/specs/otel/context/api-propagators/#b3-requirements) treat B3's parent and span information very differently. Supporting multiple tracing strategies in a single plugin (to replace otel and zipkin) might be feasible but I would consider it out of scope for this task, the propagation module could be reused in such a "common" plugin anyway.

2. that scenario is accepted and the `clear` option lets you decide what information the second plugin will receive. If both plugins execute on the request, they both affect what headers are effectively sent to the upstream. There is a test that emulates this, see `spec/02-integration/14-tracing/02-propagation_spec.lua`.

Reply:
I'm just wondering if the order matters: what if OTEL wants to set a w3 header and ZIPKIN wants to clear it?